File: scripts/scenarios/scenarios/33_tortoise_consumer.py
"""Scenario 33: "The Tortoise Consumer" — Firehose Backpressure

Verify firehose backpressure / slow consumer disconnection.
Connect to com.atproto.sync.subscribeRepos via a raw TCP socket, read the
HTTP upgrade response, then stop reading. Generate enough traffic to fill
the PDS's TCP send buffer. The PDS will eventually detect the slow consumer
and close the connection (either via backpressure check or heartbeat timeout).

Services: PDS
"""

from __future__ import annotations

import logging
import socket
import sys
import time
from pathlib import Path

_project_root = str(Path(__file__).resolve().parent.parent.parent.parent)
if _project_root not in sys.path:
    sys.path.insert(0, _project_root)

from scripts.lib.atproto import XrpcClient, get_character, PDS1, ScenarioResult, timed_call

logger = logging.getLogger("atproto.scenario")


def _connect_raw_ws(host: str, port: int, path: str) -> socket.socket:
    """Open a raw TCP socket, perform HTTP upgrade to WebSocket, and return
    the connected socket. Does NOT read any WebSocket frames after upgrade."""
    import hashlib
    import base64

    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.settimeout(10)
    sock.connect((host, port))

    # Generate WebSocket upgrade request
    key = base64.b64encode(bytes([i for i in range(16)])).decode()
    request = (
        f"GET {path} HTTP/1.1\r\n"
        f"Host: {host}:{port}\r\n"
        f"Upgrade: websocket\r\n"
        f"Connection: Upgrade\r\n"
        f"Sec-WebSocket-Key: {key}\r\n"
        f"Sec-WebSocket-Version: 13\r\n"
        f"\r\n"
    )
    sock.sendall(request.encode())

    # Read HTTP upgrade response
    response = b""
    while b"\r\n\r\n" not in response:
        chunk = sock.recv(4096)
        if not chunk:
            raise ConnectionError("Connection closed during upgrade")
        response += chunk

    if b"101" not in response.split(b"\r\n")[0]:
        raise ConnectionError(f"Upgrade failed: {response.decode(errors='replace')}")

    return sock


def run() -> ScenarioResult:
    result = ScenarioResult("Firehose Backpressure (Tortoise Consumer)")
    result.start()

    client = XrpcClient(PDS1)
    timed_call(result, "Server health check", lambda: client.wait_for_healthy(timeout=30))
    if result.failed > 0:
        result.finish()
        return result

    volt = get_character("volt")
    session = timed_call(
        result, f"Create account: {volt.handle}",
        lambda: client.accounts.create_account(volt.handle, volt.email, volt.password),
    )
    if not session:
        result.finish()
        return result
    volt.did = session["did"]
    volt.access_jwt = session["accessJwt"]

    # Parse PDS URL to get host and port
    from urllib.parse import urlparse
    parsed = urlparse(PDS1)
    host = parsed.hostname or "localhost"
    port = parsed.port or 80

    # Connect to firehose via raw TCP socket
    ws_path = "/xrpc/com.atproto.sync.subscribeRepos"
    try:
        sock = _connect_raw_ws(host, port, ws_path)
        result.step_passed("Connect to firehose")
    except Exception as exc:
        result.step_failed("Connect to firehose", str(exc))
        result.finish()
        return result

    # Read a few initial frames to confirm the connection works
    sock.settimeout(5)
    initial_count = 0
    try:
        for _ in range(5):
            data = sock.recv(65536)
            if data:
                initial_count += 1
            else:
                break
    except socket.timeout:
        pass  # No more initial data

    # Now set a very small receive buffer on the socket to limit the TCP
    # receive window. This causes the PDS's TCP send buffer to fill up,
    # which makes the PDS's internal outbound queue grow, eventually
    # triggering the ConsumerTooSlow backpressure check.
    sock.setsockopt(socket.SOL_SOCKET, socket.SO_RCVBUF, 4096)

    # Stop reading from the socket entirely. The TCP receive window will
    # close, the PDS's writes will block, and the PDS will detect the
    # slow consumer.
    sock.settimeout(120)  # Long timeout to wait for server-initiated close

    # Generate records to fill the PDS's outbound queue
    POST_COUNT = 600
    print(f"Generating {POST_COUNT} posts to fill server outbound queue...")
    for i in range(POST_COUNT):
        try:
            client.records.create_record(
                volt.did,
                "app.bsky.feed.post",
                {
                    "$type": "app.bsky.feed.post",
                    "text": f"backpressure test {i}",
                    "createdAt": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
                },
                volt.access_jwt,
            )
        except Exception as exc:
            logger.debug("Record creation %d failed: %s", i, exc)
            pass
        
        if i > 0 and i % 100 == 0:
            print(f"  Sent {i} records...")

    # Wait for the server to close the connection
    print("Waiting for PDS to drop slow consumer...")
    close_detected = False
    close_code = None
    start_time = time.time()
    timeout = 90

    while time.time() - start_time < timeout:
        # Try to read from the socket — if the server closed it, we'll get
        # empty data or an error
        try:
            data = sock.recv(65536)
            if not data:
                # Connection closed by server
                close_detected = True
                break
        except (ConnectionResetError, BrokenPipeError, OSError):
            close_detected = True
            break
        except socket.timeout:
            continue
        except Exception:
            close_detected = True
            break

        time.sleep(0.5)

    try:
        sock.close()
    except Exception:
        pass

    if close_detected:
        # The PDS closed the connection — this is the expected behavior for
        # slow consumers. We can't easily parse the WebSocket close code from
        # a raw socket, but the disconnection itself proves backpressure works.
        result.step_passed("Firehose disconnected (slow consumer dropped)",
                           f"Server closed connection after {initial_count} initial frames")
    else:
        result.step_failed("Firehose disconnected",
                           "Connection still open or timed out")

    result.finish()
    return result


if __name__ == "__main__":
    res = run()
    res.print_summary()
    sys.exit(res.exit_code)

File: scripts/scenarios/scenarios/33_tortoise_consumer.ts
/**
 * @module scenarios/33_tortoise_consumer
 *
 * Scenario: Drops a slow firehose consumer when backpressure builds.
 *
 * Behavior:
 * - Opens a raw WebSocket to com.atproto.sync.subscribeRepos and stops
 *   reading, then generates enough repo writes to exceed the server's
 *   per-connection pending-send limits.
 * - Polls for the connection to close instead of sleeping a fixed 90
 *   seconds: the test topology (docker-compose.yml, garazyk-default.json,
 *   and binary_services.ts's "pds" case) all set
 *   PDS_FIREHOSE_MAX_PENDING_SENDS=1 / PDS_FIREHOSE_MAX_PENDING_BYTES=10000,
 *   so ConsumerTooSlow should trip once real (unread) traffic backs up the
 *   connection's outbound queue, well before a real OS TCP buffer would
 *   need to fill.
 *
 * Expectations:
 * - The server sends a #error frame naming "ConsumerTooSlow" and closes
 *   the WebSocket with code 1008, within the poll window.
 */

import { getActor, PDS1 } from "../../lib/deno/config.ts";
import { now, ScenarioResult } from "../../lib/deno/runner.ts";
export { ScenarioResult, StepResult, StepStatus } from "../../lib/deno/runner.ts";
export type { ScenarioReport } from "../../lib/deno/runner.ts";
import { XrpcClient } from "../../lib/deno/client.ts";
import { timedCall } from "../../lib/deno/runner.ts";

/** Ceiling for detecting the server-initiated close once traffic stops
 * being read. Bounded, but generous relative to how quickly a threshold
 * of 1 pending send / 10KB should trip once the client stops reading. */
const CLOSE_DEADLINE_MS = 30_000;

function hasWebSocketCloseFrame(data: Uint8Array): boolean {
  for (let i = 0; i < data.length; i++) {
    if ((data[i] & 0x0f) === 0x08) return true;
  }
  return false;
}

/** Cheap substring check on the raw (DAG-CBOR) bytes: CBOR text strings
 * encode their UTF-8 bytes verbatim, so the server's #error frame
 * ({error: "ConsumerTooSlow", ...}) contains this literal byte sequence. */
function containsConsumerTooSlowMarker(chunks: Uint8Array[]): boolean {
  const marker = new TextEncoder().encode("ConsumerTooSlow");
  const combined = new Uint8Array(chunks.reduce((n, c) => n + c.length, 0));
  let offset = 0;
  for (const c of chunks) {
    combined.set(c, offset);
    offset += c.length;
  }
  outer: for (let i = 0; i + marker.length <= combined.length; i++) {
    for (let j = 0; j < marker.length; j++) {
      if (combined[i + j] !== marker[j]) continue outer;
    }
    return true;
  }
  return false;
}

async function connectRawWs(url: string) {
  const parsed = new URL(url);
  const host = parsed.hostname;
  const port = parseInt(parsed.port || "80");

  const conn = await Deno.connect({ hostname: host, port });
  const encoder = new TextEncoder();

  const key = btoa(String.fromCharCode(...crypto.getRandomValues(new Uint8Array(16))));
  const request = `GET /xrpc/com.atproto.sync.subscribeRepos HTTP/1.1\r\n` +
    `Host: ${host}:${port}\r\n` +
    `Upgrade: websocket\r\n` +
    `Connection: Upgrade\r\n` +
    `Sec-WebSocket-Key: ${key}\r\n` +
    `Sec-WebSocket-Version: 13\r\n\r\n`;

  await conn.write(encoder.encode(request));

  // Read upgrade response
  const buffer = new Uint8Array(4096);
  const n = await conn.read(buffer);
  const response = new TextDecoder().decode(buffer.subarray(0, n || 0));

  if (!response.includes("101")) {
    conn.close();
    throw new Error(`Upgrade failed: ${response}`);
  }

  return conn;
}

/**
 * Read from `conn` in a tight loop until a close is observed or
 * `deadlineMs` elapses, returning as soon as either happens (no fixed
 * sleep). A single background timer forces the connection closed at the
 * deadline so the in-flight `read()` unblocks instead of leaking.
 */
async function readUntilClosedOrDeadline(
  conn: Deno.Conn,
  deadlineMs: number,
): Promise<{ closed: boolean; chunks: Uint8Array[] }> {
  const chunks: Uint8Array[] = [];
  let closed = false;
  let timedOut = false;
  const timer = setTimeout(() => {
    timedOut = true;
    try {
      conn.close();
    } catch { /* already closed */ }
  }, deadlineMs);

  const buf = new Uint8Array(4096);
  try {
    while (!closed && !timedOut) {
      const n = await conn.read(buf);
      if (n === null || n === 0 || n === undefined) {
        closed = true;
        break;
      }
      const chunk = buf.subarray(0, n).slice();
      chunks.push(chunk);
      if (hasWebSocketCloseFrame(chunk)) {
        closed = true;
        break;
      }
    }
  } catch {
    // A read error after our own forced close (timeout) is expected and
    // not itself evidence of the server closing; only count it as
    // "closed" if the deadline hadn't already fired.
    if (!timedOut) closed = true;
  } finally {
    clearTimeout(timer);
  }
  return { closed: closed && !timedOut, chunks };
}

export async function run(): Promise<ScenarioResult> {
  const result = new ScenarioResult("Firehose Backpressure (Tortoise Consumer)");
  result.start();

  const client = new XrpcClient(PDS1);
  await timedCall(result, "Server health check", async () => {
    await client.waitForHealthy(30);
  });

  if (result.failed > 0) return result;

  const volt = getActor("volt");
  const session = await timedCall(result, "Create account: volt", async () => {
    return await client.accounts.createAccount(volt.handle, volt.email, volt.password);
  });

  if (!session) {
    result.finish();
    return result;
  }
  volt.did = session.did;
  volt.accessJwt = session.accessJwt;

  let conn: Deno.Conn;
  try {
    conn = await connectRawWs(PDS1);
    result.stepPassed("Connect to firehose");
  } catch (e) {
    result.stepFailed("Connect to firehose", String(e));
    result.finish();
    return result;
  }

  // Read a few initial bytes to confirm traffic
  const buf = new Uint8Array(1024);
  await conn.read(buf);

  // Now stop reading and generate traffic. Do not read again until every
  // post is sent — reading here would drain the connection's outbound
  // queue and prevent the backlog the test depends on from ever forming.
  const POST_COUNT = 600;
  console.log(`Generating ${POST_COUNT} posts...`);
  for (let i = 0; i < POST_COUNT; i++) {
    try {
      await client.records.createRecord(volt.did, "app.bsky.feed.post", {
        $type: "app.bsky.feed.post",
        text: `backpressure test ${i}`,
        createdAt: now(),
      }, volt.accessJwt);
    } catch { /* ignore */ }
    if (i % 100 === 0) console.log(`  Sent ${i} records...`);
  }

  console.log("Waiting for the server to drop the slow consumer...");
  const { closed, chunks } = await readUntilClosedOrDeadline(conn, CLOSE_DEADLINE_MS);

  if (closed) {
    result.stepPassed("Firehose disconnected (slow consumer dropped)");
    // Best-effort only: WebSocketConnection.closeWithCode: clears the
    // outbound queue before writing the close frame (see
    // WebSocketConnection.m), so the #error frame that
    // sendErrorFrameWithCode: just enqueued is racing its own close and is
    // not reliably flushed before the socket drops — often the client
    // observes nothing but the raw disconnect, which is still the
    // functionally important behavior (the slow consumer is gone).
    if (containsConsumerTooSlowMarker(chunks)) {
      result.stepPassed("Server sent a ConsumerTooSlow error frame before closing");
    } else {
      console.log(
        "[INFO] No ConsumerTooSlow marker observed before the close (race between " +
          "the error frame and closeWithCode: clearing the queue — not a scenario failure)",
      );
    }
  } else {
    result.stepFailed(
      "Firehose disconnected",
      `Connection still open after ${CLOSE_DEADLINE_MS}ms poll window`,
    );
  }

  try {
    conn.close();
  } catch { /* ignore */ }
  result.finish();
  return result;
}

if (import.meta.main) {
  run().then((res) => {
    console.log(res.summary());
    Deno.exit(res.ok ? 0 : 1);
  });
}
